### PR TITLE
Feat: New API functions: list_workspaces and list_projects

### PR DESCRIPTION
**Pull request by SebastianMorawiec**
# The problem

User could not list workspaces and projects lists:
https://zapatacomputing.atlassian.net/browse/ORQSDK-829

# This PR's solution

Create new API functions that return all workspaces and projects 

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`
- [x] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

### CODE REVIEW

File: src/orquestra/sdk/_base/_api/_wf_run.py [hunk 1]
```diff
@@ -513,6 +516,54 @@ def list_workflow_runs(
...
+
+    workspace_id_resolved = (
+        workspace_id.workspace_id
+        if isinstance(workspace_id, WorkspaceDef)
+        else workspace_id
+    )
+
+    runtime = resolved_config._get_runtime()
+
+    return runtime.list_projects(workspace_id_resolved)
```
Comment (BenjaminMummery) [inline src/orquestra/sdk/_base/_api/_wf_run.py:564]: Non-blocking suggestion: How would you feel about `resolved_workspace_id` for consistency with `resolved_config`?

```suggestion
    resolved_workspace_id: str

    resolved_workspace_id = (
        workspace_id.workspace_id
        if isinstance(workspace_id, WorkspaceDef)
        else workspace_id
    )

    runtime = resolved_config._get_runtime()

    return runtime.list_projects(resolved_workspace_id)
```

File: src/orquestra/sdk/_base/_driver/_ce_runtime.py [hunk 2]
```diff
@@ -446,3 +448,37 @@ def get_workflow_logs(
...
+    def list_workspaces(self):
+        try:
+            workspaces = self._client.list_workspaces()
+            return [
+                WorkspaceDef(workspace_id=ws.id, name=ws.displayName)
+                for ws in workspaces
+            ]
+        except (_exceptions.InvalidTokenError, _exceptions.ForbiddenError) as e:
+            raise exceptions.UnauthorizedError(
+                "Could not workspace list "
```
Comment (BenjaminMummery) [inline src/orquestra/sdk/_base/_driver/_ce_runtime.py:461]:
```suggestion
                "Could not list workspaces "
```

File: src/orquestra/sdk/schema/workflow_run.py [hunk 3]
```diff
@@ -71,3 +71,14 @@ class WorkflowRun(WorkflowRunMinimal):
 class ProjectRef(BaseModel):
     workspace_id: WorkspaceId
     project_id: ProjectId
+
+
+class WorkspaceDef(BaseModel):
```
Comment (alexjuda) [inline src/orquestra/sdk/schema/workflow_run.py:76]: As previously discussed in https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/109 `WorkspaceDef` and `ProjectDef` might not be the best names for these records. Usually we keep long-term support for types in `orquestra.sdk.schema` so after we make our mind it's gonna be set in stone. Let's catch up to figure out alternative naming together.

File: src/orquestra/sdk/_base/_driver/_client.py [hunk 4]
```diff
@@ -682,3 +685,51 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
...
+
+    def list_projects(self, workspace_id):
+        """
+        Gets the list of all projects in given workspace
+        """
+        default_tenant_id = 0
+        special_workspace = "system"
+        zri_type = "resource_group"
+
+        # we have to build project ZRI from some hardcoded values + workspaceId
```
Comment (BenjaminMummery) [inline src/orquestra/sdk/_base/_driver/_client.py:715]: Query: is there a documented spec this is based on that we could link to here?

**Review by BenjaminMummery — CHANGES_REQUESTED** [review]
I have a couple of non-functional suggestions but otherwise LGTM. :) How would you feel about  adding a tutorial for the new features?

**Review by alexjuda — CHANGES_REQUESTED** [review]

File: src/orquestra/sdk/_base/_spaces/_api.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import typing as t
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_spaces/_api.py:1]: **nitpick**: please add copyright header

File: src/orquestra/sdk/_base/_spaces/_structs.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+from dataclasses import dataclass
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_spaces/_structs.py:1]: **nitpick**: please add copyright header

**Review by alexjuda — APPROVED** [review]
LGTM! Added nitpick about copyright notice

**Review by BenjaminMummery — APPROVED** [review]

**Pull request merged**